backends/rapidpro/msg.go

```diff
@@ -401,6 +401,52 @@ func clearMsgSeen(rc redis.Conn, msg *DBMsg) {
 	rc.Do("hdel", prevWindowKey, urnFingerprint)
 }
 
+var luaExternalIDSeen = redis.NewScript(3, `-- KEYS: [Window, PrevWindow, ExternalID]
+	-- try to look up in window
+	local found = redis.call("hget", KEYS[1], KEYS[3])
+
+	-- didn't find it, try in our previous window
+	if not found then
+		found = redis.call("hget", KEYS[2], KEYS[3])
+	end
+
+	-- return the fingerprint found
+	return found
+`)
+
+func checkExternalIDSeen(b *backend, externalID string) bool {
+	r := b.redisPool.Get()
+	defer r.Close()
+
+	now := time.Now().In(time.UTC)
+	prev := now.Add(time.Hour * -24)
+	windowKey := fmt.Sprintf("seen:externalid:%s", now.Format("2006-01-02"))
+	prevWindowKey := fmt.Sprintf("seen:externalid:%s", prev.Format("2006-01-02"))
+
+	// see if there were any messages received in the past 24 hours
+	found, _ := redis.String(luaExternalIDSeen.Do(r, windowKey, prevWindowKey, externalID))
+
+	if found != "" {
+		return true
+	}
+	return false
+}
+
+var luaWriteExternalIDSeen = redis.NewScript(3, `-- KEYS: [Window, ExternalID, Seen]
+	redis.call("hset", KEYS[1], KEYS[2], KEYS[3])
+	redis.call("expire", KEYS[1], 86400)
+`)
+
+func writeExternalIDSeen(b *backend, externalID string) {
+	r := b.redisPool.Get()
+	defer r.Close()
+
+	now := time.Now().In(time.UTC)
+	windowKey := fmt.Sprintf("seen:externalid:%s", now.Format("2006-01-02"))
+
+	luaWriteExternalIDSeen.Do(r, windowKey, externalID, true)
```

Comment:
I think here it is important we use 
* the `msg.urnFingerprint()` then append the external ID to it 
* the window key should be more precise than the date alone
* we need to be able to not ignore the request if the message text is different even if for any reason the external ID is the same

Reply:
Honestly I think what's here is sufficient. This is only used by channels that are asserting that external ids are canonical so they can make the call that even if the msg body is different then it is fine.

Do agree however that they should probably be keyspaced to the channel, so the set you are adding to should use the channel UUID. But otherwise I think this is ok.

Reply:
Also we should be putting the UUID of the message in the map so that we keep the contract of returning the right UUID in our response. (same as checkMsgSeen)

Reply:
@norkans7 for `the window key should be more precise than the date alone`. I'm not sure how to go about this for a 24h window. 

If I include the hour in the key then I'll end up with more than two buckets, any suggestion on how you would approach this?

Reply:
@erikh360 never mind on the window, the date is enough

Reply:
@norkans7 Thanks, will you re-review my latest please?